### ISSUE TITLE
Understand periods of scheduler doing nothing

### ISSUE THREAD

**Issue by wojtek-t** [kubernetes/kubernetes#24408]
I'm running some experiments on scheduler code.
In particular what I'm running is:
- I'm starting 1000-node Kubemark cluster
- Killing scheduler job
- starting density test and wait until enough pods are created (and waiting for scheduling)
- I'm restarting scheduler (with increased qps limit)

I added some logs to understand what is expensive in scheduler and should be optimized - see:
https://github.com/kubernetes/kubernetes/compare/master...wojtek-t:test?expand=1

What I'm seeing are some period where scheduler seem to be doing nothing. For example see logs:

```
E0418 15:46:51.086081    4064 generic_scheduler.go:86] AAA -> compute predicates
E0418 15:46:51.092237    4064 generic_scheduler.go:100] AAA -> prioritize
E0418 15:46:51.097980    4064 generic_scheduler.go:237]   -> 4
E0418 15:46:51.098422    4064 node_affinity.go:96]   -> node affinity
E0418 15:46:51.100128    4064 priorities.go:92]   -> least requested
E0418 15:46:51.100347    4064 priorities.go:229]   -> balanced resource
E0418 15:46:51.104830    4064 selector_spreading.go:236]   -> spreading
E0418 15:46:51.105064    4064 generic_scheduler.go:269] AAA -> finished prioritization
```

^^ there are 5ms between "prioritize" and "4", and from the code it seems that there isn't anything happening then:
https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/scheduler/generic_scheduler.go#L98

This seems to be a very common patter in logs:

```
E0418 15:46:51.229580    4064 generic_scheduler.go:86] AAA -> compute predicates
E0418 15:46:51.235562    4064 generic_scheduler.go:100] AAA -> prioritize
E0418 15:46:51.242312    4064 generic_scheduler.go:237]   -> 4
E0418 15:46:51.242748    4064 node_affinity.go:96]   -> node affinity
E0418 15:46:51.244354    4064 priorities.go:92]   -> least requested
E0418 15:46:51.244551    4064 priorities.go:229]   -> balanced resource
E0418 15:46:51.249022    4064 selector_spreading.go:236]   -> spreading
```

Another similar "pause period" happens frequently after all priority function are already finished:

```
E0418 15:46:48.522108    4064 generic_scheduler.go:237]   -> 4
E0418 15:46:48.522667    4064 node_affinity.go:96]   -> node affinity
E0418 15:46:48.524575    4064 priorities.go:92]   -> least requested
E0418 15:46:48.524576    4064 priorities.go:229]   -> balanced resource
E0418 15:46:48.531036    4064 selector_spreading.go:236]   -> spreading
E0418 15:46:48.538735    4064 generic_scheduler.go:269] AAA -> finished prioritization
```

^^ all priority functions finished and still it took 7ms to "finish prioritization"
https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/scheduler/generic_scheduler.go#L262
[Theoretically this 7ms includes time to update combinedScore slice, but in many cases it takes <0.5ms, so it's most probably not that.]

For completeness - we are using Go 1.6., so IIUC this shouldn't be garbage collector.

@davidopp @gmarek @hongchaodeng @kubernetes/sig-scalability

**Comment by wojtek-t** [comment]
If anyone has any thoughs about it - let me know.

**Comment by wojtek-t** [comment]
And one more important point - master components are running on 32-core machine. Utilization is at the level of 50%, so I'm pretty sure it's not the effect of cpu-starvation.

**Comment by xiang90** [comment]
@wojtek-t Have you tried to play with https://github.com/kubernetes/kubernetes/tree/master/test/component/scheduler/perf? I found this issue too when I was playing around with scheduler. And I also found increasing the parallelism does not really help. The CPU utilization just stop at somewhere. I feel there is a contention somewhere. We can start to look into scheduler again if you want. But do you have a throughput goal in mind that we want to optimize for?

**Comment by wojtek-t** [comment]
@xiang90 - I'm going. But this issue is not about lock contention - especially for the first one there are no lock in between these 5ms. It has to be something different.

We would like to taget 100pods/s (not necessarily for 1.3, but we would like to go as high as possible).

**Comment by xiang90** [comment]
@wojtek-t OK, great! for 100 pods/s rate, what kind of schedule load should we target at? the density test workload?

**Comment by wojtek-t** [comment]
@xiang90 - we can start with that. It's pretty simple, but it's a good starting point.

**Comment by wojtek-t** [comment]
@xiang90 - after some fixes:
https://github.com/kubernetes/kubernetes/pull/24466

I'm trying to understand why the throughput is so different than on real cluster there.

One difference is that there are no ReplicationControllers there (in real density tests we have) - I will fix that. But it seems there are other differences too, that I'm trying to understand.

**Comment by xiang90** [comment]
@wojtek-t There is no spreading if I remember correctly.

**Comment by wojtek-t** [comment]
I checked it - there is spreading, but if there aren't any RCs and Services, it just immediately exists.

**Comment by xiang90** [comment]
@wojtek-t You are more accurate. Sorry for the confusion.

**Comment by wojtek-t** [comment]
That said - it has no effect. So what we should do is to actually add RC to the test. I will add it tomorrow, potentially with other fixes when I find something (I mean there has to be something, because there is 35 vs 120 throughput difference :P)

**Comment by xiang90** [comment]
@wojtek-t How did you get the 35? Density test without any rate limiting?

**Comment by wojtek-t** [comment]
@xiang90 - see the beginning of my first post:
https://github.com/kubernetes/kubernetes/issues/24408#issue-149197044

It is something between 35 and 40 at the beginning. But it's 3-4x smaller than what we get from benchmark. we would like to have this roughly equal to make the benchmark representative.

**Comment by xiang90** [comment]
@wojtek-t From our kubemark result, currently the bottleneck is not scheduling for default setting. The scheduling rate matches the creating rate and there is almost no waiting pods. So we need to create faster.

https://00e9e64bacbe98e31d883d9617394a6b645bb38bc598d956c1-apidata.googleusercontent.com/download/storage/v1_internal/b/metrics-kscale/o/results%2F2016-04-19%2Fkubemark%2Fdensity-test-1knodes%2Fdensity-all.svg?qk=AD5uMEs1Mj8uLkY-ytUSQdMBIN4dfGyqcbwPRpzY9DoblwneU5foaJDKBq9IxiL9NhEnvszoJmwJfXBTljpRBgRtqfpirEZXGcvSHCuosep4rrFsCTH10mm-icqMCy2BxVq9vdS6XwEty_1p5b3pRLTthxqDqxm5wmi0nZNTQaXHjOVcH0_9oNQG8-Hf83C7ZFE6d58P9hpm2yS3zN5QrYTL4clXTWTUgDaMoErFaxFGxF024RioomMeVbM6DCS1-7B-9virwQLI9NmVwdrYj0tG2DWCAuL1PW2oLM-vsoLFI2XgBbCPRj59XfgiuD0pLdZkh9NoFFLd3YtJzegSCxOBdB2rKLdK6IHERiER9oCRFCpSJEOpPrcAeMO_MuiZMSCDbGhcdmu-TYdDQ1JIy_iPQ8f0v1gE8ST5236TBzd2uI8FEmKgc0rfXKr6YXvGTpXWy4lFNMNA3N2G1layrDr306vtjjMbEuJfjdFGUwzn8r7Nm7gNwluKi34x33Rdc_OpKdnwTDqIgEgffgpFiIrTsAdIzvWskwrnxPG4J4zUoXcTlBrVh0fU1x6jTUTKyiv7dRUZqg_iimS7hpZxE2WVcqJ7GNkw8bpOF9-w2xzaUj0BOTw6VrlmqmH4rZxZvfo8H-b_gk4k3VnxRMhV1-17IhC_LbT_Dli7ufnHAF2LrH99umamwE3ccfqy507ktbKiefmVEwl4YfcIH-hGOgaSfdqaAT5mKuiymuCteOlrL1_QGFiu8-JhAF7Tbkm9E--Ms-EYP4J6NE09xcypEn8oYpub8FgTq4BB2nzupgumwsNZcndXIQavLtXqi2nxh6Dnk-7XAv3AKousznkt15I80BSnaepNKQBwMqsBAGt9GToJu1LC10rBSAbFxBeyNaYQXClfJ_hoaofTV7i2WRXLq8uDRFnT9g

**Comment by xiang90** [comment]
@wojtek-t This is when scheduling is the bottleneck: https://coreos.com/assets/images/media/scale-02.png

**Comment by wojtek-t** [comment]
@xiang90 - yes that's correct. And I would like to increase QPS limits in replication controller as soon as we have support for protobufs. And then it would become a problem.
So yes - it's solving the problem that is not yet uncovered in our tests :)

**Comment by wojtek-t** [comment]
@xiang90 - so yes, you are right. It's not an issue for now in 1000-node cluster, but it's getting worse also if we increase the number of nodes.

**Comment by xiang90** [comment]
@wojtek-t 

> It's not an issue for now in 1000-node cluster, but it's getting worse also if we increase the number of nodes.

I agree. 

I was trying to think of the reason for the difference you saw in the benchmark and kubemark density test.

For the benchmark test, I "feel" we should make it 200 pods/second sustainably to achieve 100 pods/second actual rate to be safe. And I think there is still a lot of improvement space. The scheduler cannot use all the CPU resource as you observed (actually not even close to all based on previous test...).

**Comment by gmarek** [comment]
Actually the throughput of the scheduler in benchmark should be close to the real one, especially as we're not starved on CPU on master machines.

**Comment by wojtek-t** [comment]
Yes - the first point is to actually understand the difference. If we don't understand it, it is mostly useless.

**Comment by hongchaodeng** [comment]
xref: https://github.com/kubernetes/kubernetes/issues/23263

Hope the info from above issue could help. I will submit a fix shortly.

**Comment by hongchaodeng** [comment]
I have made it work with replication controller. The RC is simple, but basically we need to start controller manager which wasn't added before. Need some cleanup.

I need to deal with other stuff and hopefully get this done no later than tomorrow.

**Comment by wojtek-t** [comment]
@hongchaodeng - I don't want to start controller manager - introducing more dependencies makes it no longer a benchmark.

What I would like to have is to just create "ReplicationController" object.

Please don't do it - I already wrote I will do it and have it almost working now.

**Comment by hongchaodeng** [comment]
@wojtek-t Sounds good :)

**Comment by wojtek-t** [comment]
It seems that the original issue is cause by contention on logging. When I removed logs that I added and changed them to a single trace log at the end, the throughput was increased by ~25%.

**Comment by wojtek-t** [comment]
@hongchaodeng @xiang90 @gmarek - I have a strong hypothesis, that differences between benchmark and real cluster are because of lack of nodes & pods updates done by other components. In particular we are missing node update for every node once per 10s, and 2 updates of pod status done by kubelet when starting the pods.
If any of you can add this, it would be super useful, as tose can cause lock contention in scheduler which is reducing throughput.

At the same time, I found few things that are easy to improve in scheduler - I will be working on it tomorrow and hopefully will send out a PR tomorrow too.

**Comment by xiang90** [comment]
@wojtek-t 

> 2. However, when I'm running Kubemark (1000nodes, 3000pods, 1RC) - so the same configuration [I doing it as described here: #24408 (comment) (so first stopping scheduler, then starting the test and waiting until enough pods are created, and then starting scheduler back), scheduler throughput is around 35-40

Can you give me the output of the density test on kubemark? I would like to see if I can find anything.

**Comment by wojtek-t** [comment]
@xiang90  I don't have access to it now - can send you tomorrow

**Comment by xiang90** [comment]
@wojtek-t np. thanks.

**Comment by timothysc** [comment]
TL;DR, where are we at with this issue?

**Comment by wojtek-t** [comment]
@timothysc - the title is misleading, we fixed/understood the problems of "periods when scheduler wasn't doing anything".

So this issue is mostly about ensuring that scheduler benchmark and real cluster generate roughly the same results. It wasn't like that in the past, but from experiments @gmarek was doing it seemed much better recently. So we should verify it and potentially close it.

**Comment by timothysc** [comment]
So I'm going to close this one, and we'll rally on new throughput benchmarks.
